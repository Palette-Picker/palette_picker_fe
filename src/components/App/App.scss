@import './src/variables.scss';

.App {
  height: 100vh;
  color: $dark-color;
  background-color: $extra-light-color;
  font-family: $basic-font;
  text-align: center;
}


.EditPaletteModal {
  align-items: center;
  background: #ffffff;
  display: flex;
  flex-direction: column;
  font-family: $basic-font;
  height: 35%;
  justify-content: center;
  left: 30%;
  padding: 10px;
  position: fixed;
  text-align: center;
  top: 35%;
  width: 40%;
}

h2 {
  width: 80%;
}

.section--btn-holder {
  display: flex;
  justify-content: space-between;
  width: 100%;
  button {
    background-color: $extra-light-color;
    border: 1px $dark-color solid;
    border-radius: 5px;
    color: $dark-color;
    font-family: $basic-font;
    font-size: 1em;
    height: 30px;
    width: 32%;
    &:hover {
      color: $extra-light-color;
      cursor: pointer;
    }
  }
}

.modal-link {
  color: $dark-color;
  text-decoration: none;
  width: 100%;
  &:hover {
    background-color: $highlight-color;
    color: $extra-light-color;
  }
}

.btn--confirm-edit:hover {
  background-color: $highlight-color;
  color: white;
}

.btn--cancel-edit:hover {
  background-color: $dark-color;
}

.btn--delete-palette:hover {
  background-color: $error-color;
}