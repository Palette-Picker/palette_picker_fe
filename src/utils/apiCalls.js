const baseUrl = 'https://palette-picker-1906-be.herokuapp.com/api/v1';
// const baseUrl = 'http://localhost:3001/api/v1';

export const getProjects = async () => {
  const response = await fetch(`${baseUrl}/projects`)

  if (!response.ok) {
    throw Error('Unable to get projects. Try again later.')
  }
  const projects = await response.json();
  return projects;
};

export const addProject = async (newProject) => {
  const options = {
    method: 'POST',
    body: JSON.stringify({
      name: newProject
    }),
    headers: {
      'content-type': 'application/json'
    }
  };
  const response = await fetch(`${baseUrl}/projects`, options);
  if (!response.ok) {
    throw Error('Unable to add project.')
  };
  const addedProject = await response.json();
  return addedProject;
};

export const addPalette = async (newPalette) => {
  const options = {
    method: 'POST', 
    body: JSON.stringify({
      name: newPalette.name,
      color1: newPalette.color1,
      color2: newPalette.color2,
      color3: newPalette.color3,
      color4: newPalette.color4,
      color5: newPalette.color5,
      project_id: newPalette.project_id
    }),
    headers: {
      'content-type': 'application/json'
    }
  };
    const response = await fetch(`${baseUrl}/palettes`, options);
    if (!response.ok) {
      throw Error('Unable to add palette.')
    };
    const addedPalette = await response.json();
    return addedPalette;
};

export const deletePalette = async (paletteId) => {
  const options = {
    method: 'DELETE',
    headers: {
      'content-type': 'application/json'
    }
  };
  const response = await fetch(`${baseUrl}/palettes/${paletteId}`, options);
  if (!response.ok) {
    throw Error('Unable to delete the palette.')
  };
  const removedId = await response.json();
  return removedId;
};

export const deleteProject = async (projectId) => {
  const options = {
    method: 'DELETE',
    headers: {
      'content-type': 'application/json'
    }
  };
  const response = await fetch(`${baseUrl}/projects/${projectId}`, options);
  if (!response.ok) {
    throw Error('Unable to delete the project and its palettes.')
  };
  const removedId = await response.json();
  return removedId;
};

export const editProject = async (projectId, newName) => {
  const options = {
    method: 'PATCH',
    body: JSON.stringify({
      name: newName
    }),
    headers: {
      'content-type': 'application/json'
    }
  };
  const response = await fetch(`${baseUrl}/projects/${projectId}`, options);
  if (!response.ok) {
    throw Error('Unable to rename the project. Try again later.')
  };
  const updatedProject = await response.json();
  return updatedProject;
};

export const editPalette = async (changedPalette) => {
  const { id, name, color1, color2, color3, color4, color5, project_id } = changedPalette;
  const options = {
    method: 'PATCH',
    body: JSON.stringify ({
      id,
      name,
      color1,
      color2,
      color3,
      color4,
      color5,
      project_id
    }),
    headers: {
      'content-type': 'application/json'
    }
  };
  const response = await fetch(`${baseUrl}/palettes/${id}`, options);
  if (!response.ok) {
    throw Error('Unable to edit palette. Try again later.')
  }
  const updatedPalette = await response.json()
  return updatedPalette;
};