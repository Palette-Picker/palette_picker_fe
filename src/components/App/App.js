import React, { Component } from 'react';
import { Route, Link } from 'react-router-dom';
import ReactModal from 'react-modal';
import Header from '../Header/Header';
import { getProjects, deletePalette, editProject, deleteProject } from '../../utils/apiCalls';
import PaletteForm from '../PaletteForm/PaletteForm';
import ProjectsContainer from '../ProjectsContainer/ProjectsContainer';

import './App.scss';


class App extends Component {
  constructor() {
    super();
    this.state = {
      colors: [],
      projects: [],
      paletteName: '',
      paletteId: null,
      projectName: '',
      projectId: null,
      error: '',
      modalOpen: false,
    }
  }

  async componentDidMount() {
    this.updateProjects();
    this.colorCheck();
  }

  updateProjects = async () => {
    try {
      const projects = await getProjects();
      this.setState({ projects })
    } catch ({ message }) {
      this.setState({ error: message })
    }
  }

  colorCheck = () => {
    let { colors } = this.state;
    if (colors.length < 5) {
      while (colors.length < 5) {
        colors.push(this.getRandomColor())
      }
      colors = colors.map((color, i) => {
        return { [`color${i + 1}`]: color, isLocked: false }
      })
      this.setState({ colors })
    }
  }

  getRandomColor() {
    return "#000000".replace(/0/g,() => {return (~~(Math.random()*16)).toString(16);});
  }

  updateColors = () => {
    let { colors } = this.state;
    colors = colors.map((color, i) => {
      if (color.isLocked === false) {
         return {
          [`color${i + 1}`]: this.getRandomColor(), 
          isLocked: false
        }
      } else {
        return color
      }
    })
    this.setState({ colors })
  }

  toggleLock = async (index) => {
    const { colors } = this.state;
    const updatedColors = colors.map((color, i) => {
      if (index === i){
        return { 
          [`color${i + 1}`]: color[`color${i + 1}`], 
          isLocked: !color.isLocked
        }
      } else {
        return color;
      }
    })
    await this.setState({ colors: updatedColors})
  }

  handleModal = async (e, colors, project, paletteName, paletteId) => {
    if (e.target.name === 'cancel') {
      await this.setState({
        colors: [],
        paletteName: '',
        projectId: null,
        projectName: '',
        paletteId: null
      })
      this.colorCheck();
    } else if (e.target.classList.contains('palette-color')) {
      this.passPaletteNameAndColors(colors, project, paletteName, paletteId)
    }
    this.setState({
      modalOpen: !this.state.modalOpen
    })
  }

  paletteDeleteHandler = async (e) => {
    e.preventDefault();
    e.persist();
    try {
      await deletePalette(this.state.paletteId);
      this.updateProjects();
      this.handleModal(e);
    } catch (error) {
      this.setState({ error });
    }
  }

  handleProjectDelete = async (id) => {
    try {
      await deleteProject(id);
      this.updateProjects();
    } catch (error) {
      this.setState({ error });
    }
  }

  passPaletteNameAndColors = (colors, project, paletteName, paletteId) => {
    this.setState({
      colors: colors,
      paletteName,
      projectId: project.id,
      projectName: project.name,
      paletteId
    });
  }

  handleKeyPress = (e, id, name) => {
    if (e.key === "Enter") {
      e.preventDefault();
      editProject(id, name);
      e.target.blur();
    }
  }

  render() {
    const { projects, colors, paletteId, paletteName, projectName, projectId } = this.state
    return (
      <div className='App'>
        <ReactModal
          ariaHideApp={false}
          isOpen={this.state.modalOpen}
          style={{
            overlay: {
              position: "fixed",
              top: 0,
              left: 0,
              right: 0,
              bottom: 0,
              backgroundColor: "rgba(0, 0, 0, 0.85)",
            },
            content: {}
          }}
          contentLabel="Edit Palette"
          className="EditPaletteModal"
          overlayClassName="EditPaletteOverlay"
        >
          <h2>Would you like to edit this palette?</h2>
          <section className='section--btn-holder'>
            <button className='btn--confirm-edit'>
              <Link onClick={e => this.handleModal(e)} to='/' name='edit'>Yes</Link>
            </button>
            <button className='btn--cancel-edit' name='cancel' onClick={e => this.handleModal(e)}>
              Cancel
            </button>
            <button onClick={e => this.paletteDeleteHandler(e)} className='btn--delete-palette'>
              Delete Palette
            </button>
          </section>
        </ReactModal>       
        <Header />
        <main>
          <Route exact path='/' render={() => <PaletteForm
            colors={colors}
            passPaletteNameAndColors={this.passPaletteNameAndColors}
            updateColors={this.updateColors}
            toggleLock={this.toggleLock}
            projects={projects}
            updateProjects={this.updateProjects}
            paletteId={paletteId}
            newPaletteName={paletteName}
            oldProjectName={projectName}
            selectedProjectId={projectId}
          />} />
          <Route exact path='/projects' render={() => <ProjectsContainer
            projects={this.state.projects}
            handleKeyPress={this.handleKeyPress}
            handleProjectDelete={this.handleProjectDelete}
            handleModal={this.handleModal} />} />
        </main>
      </div>
    )
  }
}

export default App;
