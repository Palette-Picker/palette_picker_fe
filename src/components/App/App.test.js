import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import { getProjects } from '../../utils/apiCalls';

jest.mock('../../utils/apiCalls');

describe('App', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <App />
    );
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state when colorCheck is called', () => {
    wrapper.instance().colorCheck();
    expect(wrapper.state().colors.length).toBe(5);
  });

  it('should update state when updateColors is called', () => {
    wrapper.instance().updateColors();
    expect(wrapper.state().colors.length).toBe(5);
  });

  it('should update state when toggleLock is called', () => {
    wrapper.instance().colorCheck();
    expect(wrapper.state().colors.length).toBe(5);
    wrapper.instance().toggleLock(1);
    expect(wrapper.state().colors[1].isLocked).toBe(true);
  });

  it('should reset portions of state if handleModal is called with cancel', () => {
    const mockEvent = { target: { name: 'cancel' } }
    const expected = {
      colors: [],
      error: '',
      modalOpen: false,
      paletteName: '',
      projectId: null,
      projectName: '',
      paletteId: null,
      projects: undefined
    }
    wrapper.setState({
      colors: ['color1', 'color2'],
      paletteName: 'New Palette',
      projectId: 2,
      projectName: 'Project 1',
      paletteId: 3,
    });
    wrapper.instance().handleModal(mockEvent);
    expect(wrapper.state()).toEqual(expected);
  });
})
