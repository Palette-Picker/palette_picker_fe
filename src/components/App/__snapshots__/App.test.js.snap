// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Modal
    ariaHideApp={false}
    bodyOpenClassName="ReactModal__Body--open"
    className="EditPaletteModal"
    closeTimeoutMS={0}
    contentLabel="Edit Palette"
    isOpen={false}
    overlayClassName="EditPaletteOverlay"
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {},
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.85)",
          "bottom": 0,
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
        },
      }
    }
  >
    <h2>
      Would you like to edit this palette?
    </h2>
    <section
      className="section--btn-holder"
    >
      <button
        className="btn--confirm-edit"
      >
        <Link
          className="modal-link"
          name="edit"
          onClick={[Function]}
          to="/"
        >
          Yes
        </Link>
      </button>
      <button
        className="btn--cancel-edit"
        name="cancel"
        onClick={[Function]}
      >
        Cancel
      </button>
      <button
        className="btn--delete-palette"
        onClick={[Function]}
      >
        Delete Palette
      </button>
    </section>
  </Modal>
  <Header />
  <main>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/projects"
      render={[Function]}
    />
  </main>
</div>
`;
